godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/type/array/CoreVariantArray.kt

```diff
@@ -0,0 +1,419 @@
+            Variant(
+                callNative {
+                    checkNotNull(Godot.gdnative11.godot_array_max)(it)
+                }
+            )
+        )
+    }
+
+    override fun min(): T {
+        return getCore(
+            Variant(
+                callNative {
+                    checkNotNull(Godot.gdnative11.godot_array_min)(it)
+                }
+            )
+        )
+    }
+
+    override fun popBack(): T {
+        return getCore(
+            Variant(
+                callNative {
+                    checkNotNull(Godot.gdnative.godot_array_pop_back)(it)
+                }
+            )
+        )
+    }
+
+
+    override fun popFront(): T {
+        return getCore(
+            Variant(
+                callNative {
+                    checkNotNull(Godot.gdnative.godot_array_pop_front)(it)
+                }
+            )
+        )
+    }
+
+    override fun pushBack(value: T) {
+        return callNative {
+            checkNotNull(Godot.gdnative.godot_array_push_back)(it, value.toVariant()._handle.ptr)
+        }
+    }
+
+    override fun pushFront(value: T) {
+        return callNative {
+            checkNotNull(Godot.gdnative.godot_array_push_front)(it, value.toVariant()._handle.ptr)
+        }
+    }
+
+    override fun rfind(what: T, from: Int): Int {
+        return callNative {
+            checkNotNull(Godot.gdnative.godot_array_rfind)(it, what.toVariant()._handle.ptr, from)
+        }
+    }
+
+    override fun slice(begin: Int, end: Int, step: Int, deep: Boolean): CoreVariantArray<T> {
+        return getCoreArray(
+            callNative {
+                checkNotNull(Godot.gdnative12.godot_array_slice)(it, begin, end, step, deep)
+            }
+        )
+    }
+
+    //UTILITIES
+    protected abstract fun getCore(value: Variant): T
+    protected abstract fun getCoreArray(value: CValue<godot_array>): CoreVariantArray<T>
+
+    override operator fun set(idx: Int, data: T) {
+        callNative {
+            checkNotNull(Godot.gdnative.godot_array_set)(it, idx, data.toVariant()._handle.ptr)
+        }
+    }
+
+    override operator fun get(idx: Int): T {
+        return getCore(
+            Variant(
+                callNative {
+                    checkNotNull(Godot.gdnative.godot_array_get)(it, idx)
+                }
+            )
+        )
+    }
+
+    override fun plus(other: T) {
+        this.append(other)
+    }
+
+    override fun iterator(): Iterator<T> {
+        return IndexedIterator(size(), this::get)
+    }
+}
+
+class AABBArray : CoreVariantArray<AABB> {
+    constructor() : super()
+    constructor(other: AABBArray) : super(other)
+    internal constructor(native: CValue<godot_array>) : super(native)
+    internal constructor(mem: COpaquePointer) : super(mem)
+
+    override fun getCore(value: Variant): AABB = value.asAABB()
+    override fun getCoreArray(value: CValue<godot_array>) = AABBArray(value)
+}
+
+fun AABBArrayOf(vararg elements: AABB) = AABBArray().also {
```

Comment:
I know i'll introduce a ton of work with my proposal here, but maybe it's a good idea:
How about an additional extension function like this:
`fun Iterable<AABB>.toVariantArray(): AABBArray`

Reply:
Of course we would need that for all array types...

Reply:
I'm not sure about that because that would probably something we'll have to remove for the next 4.0 version.
Today(3.2) it's fine because being the hood everything is a cpp VariantArray, it's just wrapped by a generic kotlin object. We just need to take the CValue and put it into a  different kotlin typed Array.
But soon, those two will be different objects and we'll need a wrap the elements into a Variant one by one and insert them into the VariantArray which could be slow unless GDNative provide us the method directly.

Reply:
Yes that's true but either we provide this extension function for the user or he does it by himselve when he uses kotlin iterables internally and passes them to godot. In any way there is a performance penalty. If we do it for the user or if he does it by himselve

Reply:
I agree with @chippmann .
We will have to modify a lot bindings for 4.0. But for now we're on 3.2 so we need to provide user friendly bindings for 3.2